cmd/boulder-wfe2/main.go

```diff
@@ -214,13 +237,64 @@ func loadCertificateChains(chainConfig map[string][]string, requireAtLeastOneCha
 
 		// Save the full PEM chain contents, if any
 		if buffer.Len() > 0 {
-			results[aiaIssuerURL] = buffer.Bytes()
+			results[id] = buffer.Bytes()
 		}
 	}
 
 	return results, issuerCerts, nil
 }
 
+// loadChain takes a list of filenames containing pem-formatted certificates,
+// and returns a chain representing all of those certificates in order. It
+// ensures that the resulting chain is valid. The final file is expected to be
+// a root certificate, which the chain will be verified against, but which will
+// not be included in the resulting chain.
+func loadChain(certFiles []string) (*issuance.Certificate, []byte, error) {
+	if len(certFiles) < 2 {
+		return nil, nil, errors.New(
+			"each chain must have at least two certificates: an intermediate and a root")
+	}
+
+	firstCert, err := core.LoadCert(certFiles[0])
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to load certificate: %w", err)
+	}
+
+	var buf bytes.Buffer
+	var nextCert *x509.Certificate
+
+	// Iterate over all certs except for the last, checking that their signature
+	// comes from the next cert in the list, and appending their pem to the buf.
+	for i := 0; i < len(certFiles)-1; i++ {
+		var cert *x509.Certificate
+		if nextCert != nil {
+			cert = nextCert
+		} else {
+			cert = firstCert
+		}
```

Comment:
nit: I think rather than wrangling `next` and `first` pointers, it would be a little clearer to have two loops: The first loop parses each of N `certFiles` into a slice of N `*x509.Certificate`. The second loop ranges from 0 to N-1, checking that the i'th `*x509.Certificate` is signed by the i+1th, and appending the PEM to our buffer.

Feel free to disregard, but thought I'd point out an alternate way if you want it.

Reply:
Yeah, the double-loop feels unnecessary, but these chains are going to be so short there's no reason to angle for efficiency here, and it makes the code much more readable. Done.

Reply:
I'm now curious where any optimizing compilers would look at the two separate loops and collapse them into one. Seems plausible...